astrokat/noisediode.py

```diff
@@ -9,257 +9,419 @@
+    timestamp = float(timestamp or _set_time_(lead_time))
+    user_logger.trace('TRACE: nd on at {} ({})'
+                      .format(timestamp,
+                              time.ctime(timestamp)))
 
-    return timestamp
+    true_timestamp = _switch_on_off_(kat,
+                                     timestamp,
+                                     switch=1)  # on
+
+    sleeptime = true_timestamp - time.time()
+    user_logger.debug('DEBUG: now {}, sleep {}'
+                      .format(time.time(),
+                              sleeptime))
+    time.sleep(sleeptime)  # default sleep to see for signal to get through
+    user_logger.debug('DEBUG: now {}, slept {}'
+                      .format(time.time(),
+                              sleeptime))
+    msg = ('Report: noise-diode on at {}'
+           .format(time.time()))
+    user_logger.info(msg)
+    return true_timestamp
 
 
 # switch noise-source pattern off
-def off(kat, timestamp=None, lead_time=_DEFAULT_LEAD_TIME):
+def off(kat,
+        timestamp=None,
+        lead_time=_DEFAULT_LEAD_TIME):
     """Switch noise-source pattern off.
 
     Parameters
     ----------
-    kat : Session kat container-like object
+    kat : session kat container-like object
         Container for accessing KATCP resources allocated to schedule block.
     timestamp : float, optional
         Time since the epoch as a floating point number [sec]
     lead_time : float, optional
         Lead time before the noisediode is switched off [sec]
-
     """
-    # Noise Diodes are triggered on all antennas in array simultaneously
-    # add some lead to ensure all digitisers set at the same time
-    if timestamp is None:
-        user_logger.trace(
-            "TRACE: ts + leadtime = {} + {}".format(time.time(), lead_time)
-        )
-        timestamp = np.ceil(time.time() + lead_time)
-    user_logger.trace(
-        "TRACE: nd off at {} ({})".format(timestamp, time.ctime(timestamp))
-    )
-    msg = "Request: Switch noise-diode off at {}".format(timestamp)
-    user_logger.info(msg)
-
-    # Noise Diodes are triggered on all antennas in array simultaneously
-    # add a second to ensure all digitisers set at the same time
-    replies = kat.ants.req.dig_noise_source(timestamp, 0)
-    if not kat.dry_run:
-        timestamp = _katcp_reply_to_log_(replies)
-    else:
-        # - use integer second boundary as that is most likely be an exact
-        #   time that DMC can execute at
-        msg = "Dry-run: Set all noise diodes with timestamp {} ({})".format(
-            np.ceil(timestamp), time.ctime(timestamp)
-        )
-        user_logger.info(msg)
 
-    return timestamp
+    timestamp = float(timestamp or _set_time_(lead_time))
+    user_logger.trace('TRACE: nd off at {} ({})'
+                      .format(timestamp,
+                              time.ctime(timestamp)))
+    true_timestamp = _switch_on_off_(kat, timestamp)  # default off
+    return true_timestamp
```

Comment:
The sleep in the `on` statement is to avoid you switching off the ND before it can switch on during a trigger.

Bit more explanation:
From what I understand from Marc, digitiser commands are not queued. They are executed as received. So the scenario I observed during some observations was that during a trigger event the `on` command with a timestamp was sent, but while the ND was waiting for the timestamp in order to switch on, the other command to switch `off` the ND at a timestamp arrived, which obliterated the `on` command and the ND happily waited for the `off` timestamp to switch off, even though it was never on...
Forcing this small sleep in the `on` prevents the command from being sent before the ND is ready to switch on and thus the trigger works as expected.

I have never seen an astronomer doing similar things when it comes to `off`, so the implementation is limited to `on`. At least until I have a good working example from an observation showing what needs to add for a more extensive `off` implementation.

(Please note, we have very little issues with the ND patterns, everything else are "non-standard")